### PR TITLE
Birthday

### PR DESCRIPTION
**Pull request by burnsed**
The first cut at a kinesis consumer for sending notifications to slack.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+// so that it can be directly used by the output. The returned tags will be passed along
+// with the encoded log to SendBatch()
+func (s *slackOutput) ProcessMessage(rawmsg []byte) (msg []byte, tags []string, err error) {
+	// Parse the log line
+	fields, err := decode.ParseAndEnhance(string(rawmsg), s.deployEnv, false, false, s.minTimestamp)
+	if err != nil {
+		return nil, []string{}, err
+	}
+
+	return s.encodeMessage(fields)
```
Comment (xavi-) [inline main.go]: Crazy idea: we could tag messages by username+icon.  That way the batcher could just concat all messages in a batch.  We'd also sent the max batch size to the max slack messages size.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+	icon, err := s.fetchField(fields, "_kvmeta.icon")
+	if err != nil {
+		return nil, []string{}, err
+	}
+
+	user, err := s.fetchField(fields, "_kvmeta.user")
+	if err != nil {
+		return nil, []string{}, err
+	}
```
Comment (burnsed) [inline main.go]: whoa, didn't realize that was in the decode package. I'll update!

File: main.go (hunk 1 above)
Comment (burnsed) [inline main.go]: done!

File: main.go [hunk 3]
```diff
@@ -182,23 +160,67 @@ func (s *slackOutput) ProcessMessage(rawmsg []byte) (msg []byte, tags []string,
...
+		}
+		tmpMessages = append(tmpMessages, msg)
+		length += len(msg) + 1
+	}
 
-	for _, msg := range batch {
-		err := s.sendMessage(msg, tag)
+	// Make sure the final batch gets added
+	if len(tmpMessages) > 0 {
+		texts = append(texts, strings.Join(tmpMessages, "\n"))
```
Comment (xavi-) [inline main.go]: Why not concat `texts` into one long string since they're all going to the same channel/icon/user-name?

File: main.go [hunk 4]
```diff
@@ -182,23 +160,67 @@ func (s *slackOutput) ProcessMessage(rawmsg []byte) (msg []byte, tags []string,
...
 	}
 
-	if len(failedMessages) > 0 {
+	// Report any messages that were too long
+	if len(longMessages) > 0 {
 		return kbc.PartialSendBatchError{
-			FailedMessages: failedMessages,
-			ErrMessage:     lastError.Error(),
+			FailedMessages: longMessages,
+			ErrMessage:     "Message length exceeds maximum length allowed by slack",
```
Comment (xavi-) [inline main.go]: Nice 👌

File: main.go [hunk 5]
```diff
@@ -182,23 +160,67 @@ func (s *slackOutput) ProcessMessage(rawmsg []byte) (msg []byte, tags []string,
...
+	for _, b := range batch {
+		msg := string(b)
+
+		// No single message can be longer than the max (this is a `>=` to account for the newline)
+		if len(msg) >= MaxMessageLength {
+			longMessages = append(longMessages, b)
+			continue
+		}
+
+		if length+len(msg)+1 > MaxMessageLength {
```
Comment (burnsed) [inline main.go]: doh, I forgot about `BatchSize`, good catch.

File: Makefile [hunk 6]
```diff
@@ -5,41 +5,40 @@ include golang.mk
...
-SPACE := $(EMPTY) $(EMPTY)
-JAVA_CLASS_PATH := $(subst $(SPACE),:,$(JARS_TO_DOWNLOAD))
-
-$(JARS_TO_DOWNLOAD):
-	mkdir -p `dirname $@`
-	curl -s -L -o $@ -O $(URL_PREFIX)`echo $@ | sed 's/$(JAR_DIR)\///g'`
-
-download_jars: $(JARS_TO_DOWNLOAD)
+TMP_DIR := /tmp/kinesis-notifications-consumer
+POM_ID := kinesisnotificationsconsumer
```
Comment (xavi-) [inline Makefile]: `POM_ID` doesn't seem to be used anywhere.

**Pull request merged**